### DISCUSSION TITLE
Ergonomic handle on arbitrary classes with customised initialization

### DISCUSSION THREAD

**Discussion by yoggi-yalla** [pydantic/pydantic#2543] [Question]
Hello! 

I am currently working in a project where we often use dateutil.relativedelta objects, and I'm struggling to find an ergonomic way of handling them in Pydantic. One requirement is that it should be possible to have a custom init function so that the model can be loaded from string/json. Here's what I came up with: 

```python
from pydantic import BaseModel, validator
from dateutil.relativedelta import relativedelta
from datetime import date


class Delta(BaseModel):
    __root__: relativedelta

    @validator("__root__", pre=True)
    def custom_init(cls, v):
        return relativedelta(months=3)

    class Config:
        arbitrary_types_allowed = True


class Schedule(BaseModel):
    start_date: date
    end_date: date
    frequency: Delta


schedule = Schedule.parse_raw(
    """
    {
        "start_date": "2020-01-01", 
        "end_date": "2021-06-01", 
        "frequency": "3m"
    }
    """
)

next_date = schedule.start_date + schedule.frequency.__root__  # This works
next_date = schedule.start_date + schedule.frequency  # I wish this worked
```

It kind of does the job, but it's not very neat. An alternative is to put the custom_init validator on every model that uses the frequency, but that would become a bit messy. I hope it makes sense what I'm trying to achieve, does anyone have any suggestions?

Thanks in advance!

**Comment by PrettyWood** [comment]
Hi @yoggi-yalla 
Don't you want to have a [custom data type](https://pydantic-docs.helpmanual.io/usage/types/#custom-data-types)? 🤔 
like this
```py
class Delta(relativedelta):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v):
        ...
```

**Reply by yoggi-yalla** [reply]
Ah that makes sense, thanks a bunch! I tried this a bit but didn't realize the validator classmethod was where I could put the custom init, but I see now how that makes sense. 

Quick follow-up question: do you have any tips on how to make this class json serializable? 

I was thinking to make a subclass of BaseModel with a designated json serializer and have all other models be subclasses of this instead of BaseModel, but it would have been neat if it could be specified within the class itself. I was hoping it would have sufficed to replace `__str__` or `__repr__` but that didn't do the trick. Thanks again!

**Reply by yoggi-yalla** [reply]
Subclassing both `relativedelta` and `str` seems to accomplish what I want:
```python
class Delta(relativedelta, str):
    def __init__(self,v):
        super().__init__(months=int(v[0])) # simplified...

    @classmethod
    def __get_validators__(cls):
        yield cls.validate
    
    def validate(cls, v):
        return Delta(v)

class Schedule(BaseModel):
    start_date: date
    end_date: date
    frequency: Delta
```
Now I can load it from a string such as "3m" and it will remember this string and use it when writing back to string. It feels a bit hacky, let's hope it doesn't cause issues further down the line! 😄
